src/codegen/compile.ml

```diff
@@ -3170,51 +3513,84 @@ module MakeCompact (Num : BigNumType) : BigNumType = struct
     try_unbox I32Type (fun _ -> match n with
         | 32 | 64 -> G.i Drop ^^ Bool.lit true
         | 8 | 16 ->
-          compile_bitand_const Int32.(logor 1l (shift_left minus_one (n + 1))) ^^
+          (* Please review carefully! *)
+          compile_bitand_const Int32.(logor 1l (shift_left minus_one (n + (32 - BitTagged.ubits_of Type.Int)))) ^^
```

Comment:
Regarding your comment, I believe this is correct:
- The shift serves for testing that the payload **including** the tag fits the desired bit width. This is now `n + 2` for `Type.Int`.
- The `or` with `1l` is to test the absence of the pointer skew bit, meaning that the is an unboxed big number.

But it is certainly good, if @ggreif double checks this.

Reply:
Great explanation, which make me wonder why we mask out the low bit since it must be zero in this case anyway. @ggreif any idea why?

Reply:
`if_tagged_scalar` only chooses this fast path when the bottom bit is 0, so the `logor 1l` is indeed redundant.